Cloudwatch: Add Documentation on Temporary Credentials (#75178)
grafana/grafana: docs/sources/datasources/aws-cloudwatch/aws-authentication/index.md

- `Credentials file` corresponds directly to the [SharedCredentialsProvider](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#SharedCredentialsProvider) provider in the [AWS SDK for Go](https://github.com/aws/aws-sdk-go).
  It reads the AWS shared credentials file to find a given profile.
  While `AWS SDK Default` will also find the shared credentials file, this option allows you to specify which profile to use without using environment variables.
  This option doesn't have any implicit fallbacks to other credential providers, and it fails if the credentials provided from the file aren't correct.
- `Access and secret key` corresponds to the [StaticProvider](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#StaticProvider) and uses the given access key ID and secret key to authenticate.
  This method doesn't have any fallbacks, and will fail if the provided key pair doesn't work.

> - `Grafana Assume Role` - With this auth provider option, Grafana Cloud users create an AWS IAM role that has a trust relationship with Grafana's AWS account. Grafana then uses [STS](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html) to generate temporary credentials on its behalf. Users with this option enabled no longer need to generate secret and access keys for users.

Reviewer: In this case I'd leave it in code font just to be consistent with the other list items; but there's an argument to be made that none of these list items should be in code font.
```suggestion
- `Grafana assume role` - With this auth provider option, Grafana Cloud users create an AWS IAM role that has a trust relationship with Grafana's AWS account. Grafana then uses [STS](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html) to generate temporary credentials on its behalf. Users with this option enabled no longer need to generate secret and access keys for users.
```

> - `Grafana assume role` - With this auth provider option, Grafana Cloud users create an AWS IAM role that has a trust relationship with Grafana's AWS account. Grafana then uses [STS](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html) to generate temporary credentials on its behalf. Users with this option enabled no longer need to generate secret and access keys for users.

Reviewer: ```suggestion
- `Grafana Assume Role` - With this auth provider option, Grafana Cloud users create an AWS IAM role that has a trust relationship with Grafana's AWS account. Grafana then uses [STS](https://docs.aws.amazon.com/STS/latest/APIReference/welcome.html) to generate temporary credentials on its behalf. Users with this option enabled no longer need to generate secret and access keys for users.
```

- `Workspace IAM role` corresponds to the [EC2RoleProvider](https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/ec2rolecreds/#EC2RoleProvider).
  The EC2RoleProvider pulls credentials for a role attached to the EC2 instance that Grafana runs on.
  You can also achieve this by using the authentication method AWS SDK Default, but this option is different as it doesn't have any fallbacks.
  This option is enabled by default only in Amazon Managed Grafana.

## Assume a role

{{% admonition type="note" %}}

> Assume a role is required for the `Grafana Assume Role`.

Reviewer: ```suggestion
Assume a role is required for the Grafana Assume Role.
```

For example, you may have one set of long term credentials for all of your AWS data sources. However, you want to limit the access each data source has to AWS (maybe one accesses staging data and another production data, for example). You could create separate credentials for each data source, each maintaining its own set of permissions to various resources. However, depending on how many data sources instances you have and how you've set them up, that might mean rotating and managing many different secret and access keys.

Instead, using the assume role functionality, you could have one set of AWS credentials for all of your AWS data sources that has only one permission—the permission to assume a role with STS. Then you could create a separate IAM role for each data source that specifies which permissions that data source can temporarily assume. Since IAM roles are not credentials, there's no need to rotate them and they may be easier to manage.

> The `Grafana Assume Role` also helps facilitate this. Using this role, Grafana's AWS account acts as the primary credential, having only the permission to assume roles in other accounts. You can then create IAM roles for Grafana's account to assume. For more information, refer to [Use Grafana assume role]({{< relref "#use-grafana-assume-role" >}}).

Reviewer: ```suggestion
The Grafana Assume Role also helps facilitate this. Using this role, Grafana's AWS account acts as the primary credential, having only the permission to assume roles in other accounts. You can then create IAM roles for Grafana's account to assume. For more information, refer to [Use Grafana assume role]({{< relref "#use-grafana-assume-role" >}}).
```

> If the **Assume Role ARN** field is left empty, Grafana uses the provided credentials from the selected authentication method directly, and permissions to AWS data must be attached directly to those credentials. The **Assume Role ARN** field is optional for all authentication methods except for `Grafana Assume Role`.

Reviewer: ```suggestion
If the **Assume Role ARN** field is left empty, Grafana uses the provided credentials from the selected authentication method directly, and permissions to AWS data must be attached directly to those credentials. The **Assume Role ARN** field is optional for all authentication methods except for Grafana Assume Role.
```

To disable this feature, refer to the [`assume_role_enabled` documentation][configure-grafana-assume-role-enabled].

### Use an external ID

{{% admonition type="note" %}}

> You cannot use an external ID for the `Grafana Assume Role` authentication provider.

Reviewer: ```suggestion
You cannot use an external ID for the Grafana Assume Role authentication provider.
```

{{% /admonition %}}

To assume a role in another account that was created with an external ID, specify the external ID in the **External ID** field.

For more information, refer to the [AWS documentation on external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html).

> When `Grafana Assume Role` is the selected authentication provider, Grafana is responsible for supplying and calling the external ID. It's displayed on the data source configuration page and is unique to your account. It's very important when creating an IAM role for `Grafana Assume Role` that you set a conditional that Grafana's AWS account can only assume your IAM role when a particular external ID is specified:

Reviewer: ```suggestion
When Grafana Assume Role is the selected authentication provider, Grafana is responsible for supplying and calling the external ID. It's displayed on the data source configuration page and is unique to your account. It's very important when creating an IAM role for `Grafana Assume Role` that you set a conditional that Grafana's AWS account can only assume your IAM role when a particular external ID is specified:
```

  The EC2RoleProvider pulls credentials for a role attached to the EC2 instance that Grafana runs on.
  You can also achieve this by using the authentication method AWS SDK Default, but this option is different as it doesn't have any fallbacks.
  This option is enabled by default only in Amazon Managed Grafana.

> - `Grafana Assume Role` - With this auth provider option, Grafana Cloud users create an AWS IAM role that has a trust relationship with Grafana's AWS account. Grafana then uses STS to generate temporary credentials on its behalf. Users with this option enabled no longer need to generate secret and access keys for users.

Reviewer: ```suggestion
- `Grafana Assume Role` - With this auth provider option, Grafana Cloud users create an AWS IAM role that has a trust relationship with Grafana's AWS account. Grafana then uses STS to generate temporary credentials on its behalf. Users with this option enabled no longer need to generate secrets and access keys for users.
```

If necessary, you can enable or disable them if you have server configuration access.
For more information, refer to the [`allowed_auth_providers` documentation][configure-grafana-allowed-auth-providers].

## Assume a role

{{% admonition type="note" %}}
Assume a role is required for the `Grafana Assume Role`.
{{% /admonition %}}

You can specify an IAM role to assume in the **Assume Role ARN** field.

> When this field is filled in, Grafana uses the provided credentials to perform an [sts:AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) call. In this scenario, the primary authentication method does not need permission to access CloudWatch directly, it just needs the ability to assume a role, while the role it assumes should have permission to access CloudWatch.

Reviewer: ```suggestion
When this field is filled in, Grafana uses the provided credentials to perform an [sts:AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) call. In this scenario, the primary authentication method does not need permission to access CloudWatch directly; it just needs the ability to assume a role, while the role it assumes should have permission to access CloudWatch.
```

> For example, you may have one set of long term credentials for all of your AWS data sources. However you want to limit the access each data source has to AWS (maybe one accesses staging data and another production data, for example). You could create separate credentials for each data source, each maintaining its own set of permissions to various resources. However, depending on how many data sources you have and how you've set them up, that might mean rotating/managing many different secret and access keys. 

Reviewer: ```suggestion
For example, you may have one set of long term credentials for all of your AWS data sources. However, you want to limit the access each data source has to AWS (maybe one accesses staging data and another production data, for example). You could create separate credentials for each data source, each maintaining its own set of permissions to various resources. However, depending on how many data sources you have and how you've set them up, that might mean rotating/managing many different secret and access keys. 
```

If the **Assume Role ARN** field is left empty, Grafana uses the provided credentials from the selected authentication method directly, and permissions to AWS data must be attached directly to those credentials. The **Assume Role ARN** field is optional for all authentication methods except for `Grafana Assume Role`.

To disable this feature, refer to the [`assume_role_enabled` documentation][configure-grafana-assume-role-enabled].

### Use an external ID

{{% admonition type="note" %}}
You cannot use an external ID for the `Grafana Assume Role` authentication provider.
{{% /admonition %}}

To assume a role in another account that was created with an external ID, specify the external ID in the **External ID** field.

For more information, refer to the [AWS documentation on external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html).

> When `Grafana Assume Role` is the selected authentication provider, Grafana is responsible for supplying and calling the external ID. It is displayed on the data source configuration page and is unique to your account. It's very important when creating an IAM role for `Grafana Assume Role` that you set a conditional that Grafana's AWS account can only assume your IAM role when a particular external ID is specified:

Reviewer: ```suggestion
When `Grafana Assume Role` is the selected authentication provider, Grafana is responsible for supplying and calling the external ID. It's displayed on the data source configuration page and is unique to your account. It's very important when creating an IAM role for `Grafana Assume Role` that you set a conditional that Grafana's AWS account can only assume your IAM role when a particular external ID is specified:
```

[configure-grafana-assume-role-enabled]: "/docs/grafana/ -> /docs/grafana/<GRAFANA VERSION>/setup-grafana/configure-grafana#assume_role_enabled"
[configure-grafana-assume-role-enabled]: "/docs/grafana-cloud/ -> /docs/grafana/<GRAFANA VERSION>/setup-grafana/configure-grafana#assume_role_enabled"
{{% /docs/reference %}}

## Use `Grafana Assume Role`

> > Note: Grafan Assume Role is currently in Private Preview for Grafana Cloud. To get early access this feature, reach out to Customer Support and ask for the `awsDatasourcesTempCredentials` feature toggle to be enabled on your account. It is currently only available for CloudWatch.

Reviewer: We should replace this with a private preview shortcode for the correct text

`Grafana Assume Role` lets Grafana Cloud users authenticate with AWS without having to create and maintain long term AWS Users and rotate their access and secret keys. Instead Grafana Cloud customers can create an IAM Role with permissions to access CloudWatch and a trust relationship with Grafana's AWS Account. Grafana's AWS Account will then make an STS request to AWS to create temporary credentials to access the customer's AWS data. It makes this STS request by passing along an `externalID` that is unique per cloud account, to ensure that Grafana Cloud customers can only access their own AWS data. For more information, refer to the [AWS documentation on external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html).

To use this `Grafana Assume Role`:

> 1. Put in a request to Customer Support to enable `awsDatasourcesTempCredentials`
> 2. Once the feature is enabled, create a new CloudWatch datasource (or update an existing one) and select `Grafana Assume Role` as an authentication provider
> 3. Create a new IAM role in the AWS console, and select Another AWS account as the Trusted entity.
> 4. Enter the account ID of the Grafana account that has permission to assume this role 008923505280 and check the Require external ID box.
> 5. Enter the external ID specified in the Instructions Box on the CloudWatch Datasource Configuration page in Grafana. This external ID will be unique to your grafana instance.
> 6. Attach any required permissions you would like Grafana to be able to access on your behalf (example CloudWatch Logs and CloudWatch Metrics policies)
> 7. Give the role a name and description, and click Create role.
> 8. Copy the ARN of the role you just created and paste it into the Assume Role ARN field on the CloudWatch Datasource Configuration page in Grafana.

Reviewer: ```suggestion
1. Put in a request to Customer Support to enable`awsDatasourcesTempCredentials`.
2. Once the feature is enabled, create a new CloudWatch data source (or update an existing one) and select `Grafana Assume Role` as an authentication provider.
3. Create a new IAM role in the AWS console, and select **Another AWS account** as the Trusted entity.
4. Enter the account ID of the Grafana account that has permission to assume this role 008923505280 and check the **Require external ID** box.
5. Enter the external ID specified in the instructions box on the **Settings** tab of CloudWatch data source configuration in Grafana. This external ID will be unique to your Grafana instance.
6. Attach any required permissions you would like Grafana to be able to access on your behalf (for example, CloudWatch Logs and CloudWatch Metrics policies).
7. Give the role a name and description, and click **Create role**.
8. Copy the ARN of the role you just created and paste it into the **Assume Role ARN** field on the **Settings** tab of CloudWatch data source configuration in Grafana.
```

Reviewer: Some questions:

1. In step 3 is Trusted entity a field name?
2. What's the number in step 4?
3. On step 5, is this a titled part of the UI or just a place where there are instructions?

  runAsGroup: 472
```

> ## Use `Grafana Assume Role`

Reviewer: ```suggestion
## Use Grafana Assume Role
```

{{% admonition type="note" %}}
Grafana Assume Role is currently in [private preview]({{< relref "https://grafana.com/docs/release-life-cycle" >}}) for Grafana Cloud. To get early access this feature, reach out to Customer Support and ask for the `awsDatasourcesTempCredentials` feature toggle to be enabled on your account. It is currently only available for CloudWatch.
{{% /admonition %}}

> The `Grafana Assume Role` authentication provider lets you authenticate with AWS without having to create and maintain long term AWS users or rotate their access and secret keys. Instead, you can create an IAM role that has permissions to access CloudWatch and a trust relationship with Grafana's AWS account. Grafana's AWS account then makes an STS request to AWS to create temporary credentials to access your AWS data. It makes this STS request by passing along an `externalID` that's unique per Cloud account, to ensure that Grafana Cloud users can only access their own AWS data. For more information, refer to the [AWS documentation on external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html).

Reviewer: ```suggestion
The Grafana Assume Role authentication provider lets you authenticate with AWS without having to create and maintain long term AWS users or rotate their access and secret keys. Instead, you can create an IAM role that has permissions to access CloudWatch and a trust relationship with Grafana's AWS account. Grafana's AWS account then makes an STS request to AWS to create temporary credentials to access your AWS data. It makes this STS request by passing along an `externalID` that's unique per Cloud account, to ensure that Grafana Cloud users can only access their own AWS data. For more information, refer to the [AWS documentation on external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html).
```

> To use the `Grafana Assume Role`:

Reviewer: ```suggestion
To use the Grafana Assume Role:
```

1. Put in a request to Customer Support to enable`awsDatasourcesTempCredentials`.

> 2. Once the feature is enabled, create a new CloudWatch data source (or update an existing one) and select `Grafana Assume Role` as an authentication provider.

Reviewer: ```suggestion
2. Once the feature is enabled, create a new CloudWatch data source (or update an existing one) and select **Grafana Assume Role** as an authentication provider.
```

> 3. In the AWS Console, create a new IAM role, and select **Another AWS account** as the Trusted Entity.

Reviewer: ```suggestion
3. In the AWS Console, create a new IAM role, and under **Trusted entity type**, select **Another AWS account** as the trusted Entity.
```

{{% /admonition %}}

The `Grafana Assume Role` authentication provider lets you authenticate with AWS without having to create and maintain long term AWS users or rotate their access and secret keys. Instead, you can create an IAM role that has permissions to access CloudWatch and a trust relationship with Grafana's AWS account. Grafana's AWS account then makes an STS request to AWS to create temporary credentials to access your AWS data. It makes this STS request by passing along an `externalID` that's unique per Cloud account, to ensure that Grafana Cloud users can only access their own AWS data. For more information, refer to the [AWS documentation on external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html).

To use the `Grafana Assume Role`:

1. Put in a request to Customer Support to enable`awsDatasourcesTempCredentials`.
2. Once the feature is enabled, create a new CloudWatch data source (or update an existing one) and select `Grafana Assume Role` as an authentication provider.

> 3. Create a new IAM role in the AWS console, and select **Another AWS account** as the Trusted entity.
> 4. Enter the account ID of the Grafana account that has permission to assume this role 008923505280 and check the **Require external ID** box.

Reviewer: A couple questions:
In step 3, is Trusted entity a field label?
In step 4, what is the number after the word "role"?

I've made edits based on the "Trusted entity" being a field label and the number being a mistake.

```suggestion
3. Create a new IAM role in the AWS console, and select **Another AWS account** as the **Trusted entity**.
4. Enter the account ID of the Grafana account that has permission to assume this role and check the **Require external ID** box.
```

Author: Yeah, I think I'd describe this as a label:
<img width="1026" alt="Screenshot 2023-09-27 at 2 55 58 PM" src="https://github.com/grafana/grafana/assets/6620164/520728ca-8d7c-421b-a92d-27d188540814">

and yeah sorry that last sentence got a bit lost from me I think. The number refers to our accountid. It will be the same for everyone which is why I put it in our documentation originally but I have removed it just to be a bit more future-proof in case we update it and forget to update the documentation. hopefully this is more clear!

Reviewer: Thanks for circling back! I've made a bit of an update to step and left step 4 as is!
